core/src/main/java/hudson/PluginManager.java

```diff
@@ -425,6 +430,190 @@ public void run(Reactor reactor) throws Exception {
         }});
     }
 
+    protected @Nonnull Set<String> loadPluginsFromWar(@Nonnull String fromPath) {
+        return loadPluginsFromWar(fromPath, null);
+    }
+
+    protected @Nonnull Set<String> loadPluginsFromWar(@Nonnull String fromPath, @CheckForNull FilenameFilter filter) {
+        Set<String> names = new HashSet();
+
+        ServletContext context = Jenkins.getInstance().servletContext;
+        Set<String> plugins = Util.fixNull((Set<String>) context.getResourcePaths(fromPath));
+        Set<URL> copiedPlugins = new HashSet<>();
+        Set<URL> dependencies = new HashSet<>();
+
+        for( String pluginPath : plugins) {
+            String fileName = pluginPath.substring(pluginPath.lastIndexOf('/')+1);
+            if(fileName.length()==0) {
+                // see http://www.nabble.com/404-Not-Found-error-when-clicking-on-help-td24508544.html
+                // I suspect some containers are returning directory names.
+                continue;
+            }
+            try {
+                URL url = context.getResource(pluginPath);
+                if (filter != null && url != null) {
+                    if (!filter.accept(new File(url.getFile()).getParentFile(), fileName)) {
+                        continue;
+                    }
+                }
+
+                names.add(fileName);
+                copyBundledPlugin(url, fileName);
+                copiedPlugins.add(url);
+                try {
+                    addDependencies(url, fromPath, dependencies);
+                } catch (Exception e) {
+                    LOGGER.log(Level.SEVERE, "Failed to resolve dependencies for the bundled plugin " + fileName, e);
+                }
+            } catch (IOException e) {
+                LOGGER.log(Level.SEVERE, "Failed to extract the bundled plugin "+fileName,e);
```

Comment:
NIT - pick one coding style and stick to it?  (compare with the LOGGER 4 lines above!